gcc/rust/resolve/rust-default-resolver.cc

```diff
@@ -0,0 +1,805 @@
+}
+
+void
+DefaultResolver::visit (AST::Enum &type)
+{
+  // FIXME: Do we need to scope anything by default?
+
+  auto variant_fn = [this, &type] () {
+    for (auto &variant : type.get_variants ())
+      variant->accept_vis (*this);
+  };
+
+  resolver.scoped (Rib::Kind::Item /* FIXME: Correct? */, type.get_node_id (),
+		   variant_fn, type.get_identifier ());
+}
+
+void
+DefaultResolver::visit (AST::BorrowExpr &expr)
+{
+  expr.get_borrowed_expr ()->accept_vis (*this);
+}
+
+void
+DefaultResolver::visit (AST::DereferenceExpr &expr)
+{
+  expr.get_dereferenced_expr ()->accept_vis (*this);
+}
+
+void
+DefaultResolver::visit (AST::ErrorPropagationExpr &expr)
+{
+  expr.get_propagating_expr ()->accept_vis (*this);
+}
+
+void
+DefaultResolver::visit (AST::NegationExpr &expr)
+{
+  expr.get_negated_expr ()->accept_vis (*this);
+}
+
+void
+DefaultResolver::visit (AST::ArithmeticOrLogicalExpr &expr)
+{
+  expr.get_left_expr ()->accept_vis (*this);
+  expr.get_right_expr ()->accept_vis (*this);
+}
+
+void
+DefaultResolver::visit (AST::ComparisonExpr &expr)
+{
+  expr.get_left_expr ()->accept_vis (*this);
+  expr.get_right_expr ()->accept_vis (*this);
+}
+
+void
+DefaultResolver::visit (AST::LazyBooleanExpr &expr)
+{
+  expr.get_left_expr ()->accept_vis (*this);
+  expr.get_right_expr ()->accept_vis (*this);
+}
+
+void
+DefaultResolver::visit (AST::TypeCastExpr &expr)
+{
+  expr.get_type_to_cast_to ()->accept_vis (*this);
+  expr.get_casted_expr ()->accept_vis (*this);
+}
+
+void
+DefaultResolver::visit (AST::AssignmentExpr &expr)
+{
+  expr.get_left_expr ()->accept_vis (*this);
+  expr.get_right_expr ()->accept_vis (*this);
+}
+
+void
+DefaultResolver::visit (AST::CompoundAssignmentExpr &expr)
+{
+  expr.get_left_expr ()->accept_vis (*this);
+  expr.get_right_expr ()->accept_vis (*this);
+}
+
+void
+DefaultResolver::visit (AST::GroupedExpr &expr)
+{
+  expr.get_expr_in_parens ()->accept_vis (*this);
+}
+
+void
+DefaultResolver::visit (AST::ArrayElemsValues &array)
+{
+  for (auto &value : array.get_values ())
+    value->accept_vis (*this);
+}
+
+void
+DefaultResolver::visit (AST::ArrayElemsCopied &array)
+{
+  array.get_elem_to_copy ()->accept_vis (*this);
+  array.get_num_copies ()->accept_vis (*this);
+}
+
+void
+DefaultResolver::visit (AST::ArrayExpr &expr)
+{
+  expr.get_array_elems ()->accept_vis (*this);
+}
+
+void
+DefaultResolver::visit (AST::ArrayIndexExpr &expr)
+{
+  expr.get_array_expr ()->accept_vis (*this);
+  expr.get_index_expr ()->accept_vis (*this);
+}
+
+void
+DefaultResolver::visit (AST::TupleExpr &expr)
+{
+  for (auto &element : expr.get_tuple_elems ())
+    element->accept_vis (*this);
+}
+
+void
+DefaultResolver::visit (AST::TupleIndexExpr &expr)
+{
+  expr.get_tuple_expr ()->accept_vis (*this);
+}
+
+void
+DefaultResolver::visit (AST::StructExprFieldIdentifierValue &)
+{}
+
+void
+DefaultResolver::visit (AST::StructExprFieldIndexValue &)
+{}
+
+void
+DefaultResolver::visit (AST::CallExpr &expr)
+{}
```

Comment:
Should everything be visiting children here? The classes derived from this are presumably also going to be handling uses,

Reply:
yes, this should visit the called expression as well as each given argument and generic arguments if any - good catch